daliuge-engine/dlg/apps/app_base.py

```diff
@@ -393,15 +456,10 @@ def dropCompleted(self, uid, drop_state):
                 self.async_execute()
 
     def async_execute(self):
-        # Return immediately, but schedule the execution of this app
-        # If we have been given a thread pool use that
-        if hasattr(self, "_tp"):
-            self._tp.apply_async(self._execute_and_log_exception)
-        else:
-            t = threading.Thread(target=self._execute_and_log_exception)
-            t.daemon = 1
-            t.start()
-            return t
+        # TODO Do we need another thread pool for this?
```

Comment:
That's what I initially had tried to do, but it very quickly became very complex.

I also always had it in the back of my mind that it just seems soooo wasteful to create/use a thread only to schedule the execution of `run()` (which then might end up running on a separate process). `asyncio` indeed should be the way to go, but hey, this started as a 2.7-compatible project!

Reply:
Yeah, wasteful, plus a memory leak which is the bigger issue IMO. But that seems like a issue to tackle in another MR :wink: